queue/server.go

```diff
@@ -0,0 +1,138 @@
+package queue
+
+import (
+	"sync"
+
+	"context"
+	"errors"
+	"time"
+
+	"github.com/centrifuge/gocelery"
+	logging "github.com/ipfs/go-log"
+)
+
+const (
+	BlockHeightParam string = "BlockHeight"
+	TimeoutParam     string = "Timeout"
+)
+
+var log = logging.Logger("queue-server")
+
+// Config is an interface for queue specific configurations
+type Config interface {
+
+	// GetNumWorkers gets the number of background workers to initiate
+	GetNumWorkers() int
+
+	// GetWorkerWaitTime gets the worker wait time for a task to be available while polling
+	// increasing this may slow down task execution while reducing it may consume a lot of CPU cycles
+	GetWorkerWaitTimeMS() int
+}
+
+// QueuedTaskType is a task to be queued in the centrifuge node to be completed asynchronously
+type QueuedTaskType interface {
+
+	// TaskTypeName of the task
+	TaskTypeName() string
+}
+
+// QueuedTaskResult represents a result from a queued task execution
+type QueuedTaskResult interface {
+
+	// Get the result within a timeout from the queue task execution
+	Get(timeout time.Duration) (interface{}, error)
+}
+
+// Server represents the queue server currently implemented based on gocelery
+type Server struct {
+	config    Config
+	queue     *gocelery.CeleryClient
+	taskTypes []QueuedTaskType
+	stop      chan bool
+	lock 	  sync.RWMutex
+}
+
+// TaskTypeName of the queue server
+func (qs *Server) Name() string {
+	return "QueueServer"
+}
+
+// Start the queue server
+func (qs *Server) Start(ctx context.Context, wg *sync.WaitGroup, startupErr chan<- error) {
+	defer wg.Done()
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+	var err error
+	qs.queue, err = gocelery.NewCeleryClient(
+		gocelery.NewInMemoryBroker(),
+		gocelery.NewInMemoryBackend(),
+		qs.config.GetNumWorkers(),
+		qs.config.GetWorkerWaitTimeMS(),
+	)
+	if err != nil {
+		startupErr <- err
+	}
+	for _, task := range qs.taskTypes {
+		qs.queue.Register(task.TaskTypeName(), task)
+	}
+	// start the workers
+	qs.queue.StartWorker()
+
+	qs.stop = make(chan bool)
+	for {
+		select {
+		case <-qs.stop:
```

Comment:
This would not work since select doesn't fallthrough. I would recommend using context.WithCancel and hold the cancelFunc instead of channel